### ISSUE TITLE
AmazonS3Client.CopyObject() is incorrectly reporting missing DestinationKey

### ISSUE THREAD

**Issue by mwhouser** [aws/aws-sdk-net#1720]
## Description
When copying an object using `Amazon.S3.AmazonS3Client.CopyObject`, if the `DestinationKey` is set to `"/"`, then an exception is thrown reporting that the value is not set.

## Use Case

I have an actual object "/" (without the quotes) in my bucket that I want copied.

## Reproduction Steps

```csharp
var client = new Amazon.S3.AmazonS3Client(Amazon.RegionEndpoint.USEast1);

var request = new Amazon.S3.Model.CopyObjectRequest()
{
  SourceKey = "/",
  DestinationKey = "/",

  SourceBucket = "bucket1",
  DestinationBucket = "bucket2",
};

try
{
  var response = client.CopyObject(request);
}
catch (Exception e)
{
  Console.WriteLine("{0}", e.Message);
}

Console.WriteLine("Press Enter to continue...");
Console.ReadLine();
```

The above code throws an exception and outputs the following:

```
DestinationKey is a required property and must be set before making this call.
Parameter name: CopyObjectRequest.DestinationKey
Press Enter to continue...
```

### Environment

* SDK Version: AWSSDK.S3 version 3.5.1.9
* Package Version: AWSSDK.S3 version 3.5.1.9
* OS Info: Windows 10
* Build Environment Visual Studio 2019
* Targeted .NET Platform:  4.7.2

## Additional Notes

1. This appears to only happen when the object is "/". Copying of other objects works correctly.

2. This issue occurs without a valid AWS role or key. So it appears to be an SDK validation check before the `s3:CopyObject` API is actually called.

---

This is a :bug: bug-report

**Comment by ashishdhingra** [comment]
Hi @mwhouser,

Good morning.

I don't think this is a bug. The Source and Destination keys must point to valid object. The character `/` specifies the folder structure, not the actual object. Kindly note that this is not similar to existing bug https://github.com/aws/aws-sdk-net/issues/1241 where `CopyObject fails for file starting with "/" `. 

If you want to copy the folder structure then you might have to write logic that iterates through source folder objects and copies them one by one. The [CopyObject](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CopyObject.html) API call handles single object.

Hope this answers your question.

Thanks,
Ashish

**Comment by mwhouser** [comment]
Hi @ashishdhingra 

> The Source and Destination keys must point to valid object.

It does. "/" is a valid object key.

> The character / specifies the folder structure, not the actual object. 

That is not correct. "Folder structure" is a human concept laid over S3 as an organizational system. S3 keys are just that... keys. The `/` characters in keys don't mean anything to S3 aside when used by the APIs as delimiters. And as such, I can choose to use any other character as my "Folder structure" delimiter... such as an `M` character if I so choose.

I can create objects with all 3 of the following keys and they will all be distinctly different objects:

* "abc"
* "/abc"
* "abc/"

Try it yourself with the following sample program. Also notice that I am writing to an object called "/" and reading it's contents too. So I should be able to copy that object.

```csharp
static void PutStringsDemo(Amazon.S3.IAmazonS3 client)
{
    string bucketName = "bucket1";

    PutObjectString(bucketName, "abc", "This is object 'abc'", client);
    PutObjectString(bucketName, "abc/", "This is object 'abc/'", client);
    PutObjectString(bucketName, "/abc", "This is object '/abc'", client);
    PutObjectString(bucketName, "/", "This is object '/'", client);

    Console.WriteLine("Object abc: {0}", GetObjectString(bucketName, "abc", client));
    Console.WriteLine("Object abc/: {0}", GetObjectString(bucketName, "abc/", client));
    Console.WriteLine("Object /abc: {0}", GetObjectString(bucketName, "/abc", client));
    Console.WriteLine("Object /: {0}", GetObjectString(bucketName, "/", client));
}

static void PutObjectString(string bucketName, string key, string value, Amazon.S3.IAmazonS3 client)
{
    using (var stream = new MemoryStream())
    {
        using (var writer = new StreamWriter(stream, Encoding.UTF8, 1024, true))
        {
            writer.Write("{0}", value);
            writer.Flush();
        }

        stream.Seek(0, SeekOrigin.Begin);

        var request = new Amazon.S3.Model.PutObjectRequest()
        {
            Key = key,
            BucketName = bucketName,

            InputStream = stream,
        };

        var response = client.PutObject(request);
    }
}

static string GetObjectString(string bucketName, string key, Amazon.S3.IAmazonS3 client)
{
    var request = new Amazon.S3.Model.GetObjectRequest()
    {
        Key = key,
        BucketName = bucketName,
    };

    using (var response = client.GetObject(request))
    {
        using (var stream = response.ResponseStream)
        {
            using (var reader = new StreamReader(stream))
            {
                return reader.ReadToEnd();
            }
        }
    }
}
```

When I execute this program, I get the following output:

```
Object abc: This is object 'abc'
Object abc/: This is object 'abc/'
Object /abc: This is object '/abc'
Object /: This is object '/'
```

After executing this program, I have 4 objects in my bucket, as seen by `s3:ListObjects`:

```
$ aws s3api list-objects --bucket bucket1
{
    "Contents": [
        {
            "LastModified": "2020-09-25T20:13:36.000Z",
            "ETag": "\"03f9b0ffc1d9c2286a608660775d35a4\"",
            "StorageClass": "STANDARD",
            "Key": "/",
            "Size": 21
        },
        {
            "LastModified": "2020-09-25T20:13:36.000Z",
            "ETag": "\"92f596e88248ff7f4b11d774636be935\"",
            "StorageClass": "STANDARD",
            "Key": "/abc",
            "Size": 24
        },
        {
            "LastModified": "2020-09-25T20:13:36.000Z",
            "ETag": "\"fad4fd3092dd59db8ea2fa52dbd49996\"",
            "StorageClass": "STANDARD",
            "Key": "abc",
            "Size": 23
        },
        {
            "LastModified": "2020-09-25T20:13:36.000Z",
            "ETag": "\"9917757841faaf1fff236375e59aa749\"",
            "StorageClass": "STANDARD",
            "Key": "abc/",
            "Size": 24
        }
    ]
}

```

Notice the line in the first object:

```
"Key": "/",
```

So there's now an actual object of 21 bytes called "/" in my bucket.

When I examine the bucket using the AWS Management Console, I see the following:

![image](https://user-images.githubusercontent.com/1119525/94312640-41376180-ff4b-11ea-9eab-9fc348bf89c7.png)

Notice a few things:

* There's a folder with a "blank" name. That's the "/" object.
  * Navigating into that folder will show me the "/abc" object.
* There's another folder called "abc". That's "abc/".
* There's an object called "abc"  with the same name as the "abc" folder. That's "abc".

Unfortunately, the AWS Management Console does not allow me to download the contents of "folder" objects, even though it has data.

So, to S3, the "/" characters means nothing and has no additional meaning. It's just another character in an object key.

### TL;DR

It's possible to Put and Get an object named "/". So there should be no reason why I cannot copy it to another bucket.

**Comment by mwhouser** [comment]
Hi @ashishdhingra 

One more thing. I can copy the "/" object using the follow AWS CLI command:

```bash
$ aws s3api copy-object --bucket bucket2 --key '/' --copy-source bucket1//
```

Note the double slash in the `--copy-source` parameter.

**Comment by ashishdhingra** [comment]
@mwhouser Thanks for your inputs. I would need to run some tests against the service API directly to see if it allows copying folder (key '/') from source to destination. I'm not sure if AWS CLI has any custom implementation. I will keep you posted.

**Comment by mwhouser** [comment]
Hi @ashishdhingra 

I found the problem.

Please see this sample code:

```csharp
var request = new Amazon.S3.Model.CopyObjectRequest()
{
        SourceKey = "/abc",
        DestinationKey = "/abc",

        SourceBucket = "bucket1",
        DestinationBucket = "bucket2",
};

var response = client.CopyObject(request);
```

I executed this code then I checked my target bucket. I don't see "/abc". Instead I see an object "abc". It's contents are the contents of "abc" in the source bucket. So the wrong file was copied. This is very, very bad.

Looking through the source code here on GitHub, I found the following:

https://github.com/aws/aws-sdk-net/blob/641affd215ac31454f9d034b8867a4f5adae76ef/sdk/src/Services/S3/Custom/Model/Internal/MarshallTransformations/CopyObjectRequestMarshaller.cs#L38

It looks like the SDK is intentionally stripping the leading slash from the source and target keys. Again, very, very bad.

So the problem is that:

1. The SDK is stripping the leading slash, then
2. The target key is now blank, so it's throwing the exception.

This behaviour is very, very bad. The SDK should not "fix" my keys for me. If I want to copy "/abc", then it should copy "/abc".

**Comment by ashishdhingra** [comment]
Do not close.

**Comment by ashishdhingra** [comment]
Still need to investigate.

**Comment by dscpinheiro** [comment]
Apologies for the long delay, but we just released a new version of S3 that does not remove the leading slash from object keys: https://www.nuget.org/packages/AWSSDK.S3/3.7.205.20

Changelog: https://github.com/aws/aws-sdk-net/blob/main/changelogs/SDK.CHANGELOG.2023.md#376740-2023-10-31-1833-utc

**Comment by dscpinheiro** [comment]
Update: We had to revert that S3 change as it caused an unexpected issue in our AWS Tools for PowerShell Cmdlets.

I'll discuss with the team what best approach forward is, and I'll provide an update here when I have one.

**Comment by dscpinheiro** [comment]
Another update: Starting in version `3.7.205.22` of the `AWSSDK.S3` package, there's a new option in `CopyRequest` and `CopyPart`: `DisableTrimmingLeadingSlash`; if it's set to `true` the .NET SDK will not remove leading slashes from the object keys:

```
// Destination object on S3 will be "/destkey" (and not "destkey" as before)
s3Client.CopyObject(new CopyObjectRequest
{
    SourceBucket = bucketName,
    SourceKey = "/sourceKey",
    DestinationBucket = bucketName,
    DestinationKey = "/destkey",
    
    DisableTrimmingLeadingSlash = true
})
```

Our previous approach was not acceptable as we found use cases that depended on the existing behavior (which we understand is inconsistent, but we cannot break their workflows without a major version change).
